{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7030e9ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import h5py\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "\n",
    "import seaborn as sns\n",
    "\n",
    "from scipy.stats import norm\n",
    "\n",
    "from basenji.gene import gtf_kv\n",
    "from basenji import gene as bgene\n",
    "\n",
    "import pybedtools\n",
    "\n",
    "from sklearn.metrics import average_precision_score, roc_auc_score\n",
    "\n",
    "#Helper function to make a bedtool object from the crispr dataframe\n",
    "def _make_crispr_bedt(crispr_df, allow_other_exons_to_overlap=False, window_size=2000, extra_padding=64):\n",
    "  \n",
    "    bed_lines = []\n",
    "    \n",
    "    #Loop over rows in crispr dataframe\n",
    "    for _, row in crispr_df.iterrows() :\n",
    "        \n",
    "        chrom = row['chr']\n",
    "        midpoint = (row['start'] + row['end']) // 2\n",
    "        start = midpoint - window_size // 2 - extra_padding\n",
    "        end = midpoint + window_size // 2 + extra_padding\n",
    "        \n",
    "        #Construct unique enhancer id\n",
    "        enh_id = row['chr'] + '_' + str(row['start']) + '_' + str(row['end'])\n",
    "        if allow_other_exons_to_overlap :\n",
    "            enh_id += '_' + row['gene']\n",
    "        \n",
    "        gene = row['gene']\n",
    "        \n",
    "        #Write to bed file\n",
    "        bed_lines.append('%s %d %d %s %s' % (chrom, start, end, enh_id, gene))\n",
    "    \n",
    "    #Create bedtool object\n",
    "    bedt = pybedtools.BedTool('\\n'.join(bed_lines), from_string=True)\n",
    "    \n",
    "    return bedt\n",
    "\n",
    "#Helper function to compute gradient/ISM scores for each enhancer-gene pair in the crispr dataframe\n",
    "def _score_sites_borzoi(crispr_df, gene_i, grads_ref, seq_starts, seq_ends, elen_ext=3, extra_padding=1064, normalize=True, use_abs=True, use_gaussian=False, sigma=300) :\n",
    "    \n",
    "    #Get sequence length (of attributions)\n",
    "    seq_len = grads_ref.shape[1]\n",
    "    \n",
    "    #Initialize scores and metadata\n",
    "    enh_ref = np.zeros(crispr_df.shape[0])\n",
    "    enh_seq_starts = np.zeros(crispr_df.shape[0], dtype='int32')\n",
    "    enh_seq_ends = np.zeros(crispr_df.shape[0], dtype='int32')\n",
    "    enh_valids = np.zeros(crispr_df.shape[0], dtype='int32')\n",
    "    \n",
    "    #Define smoothing helpers\n",
    "    if use_gaussian :\n",
    "        window = np.arange(-4*sigma, 4*sigma+1)\n",
    "        weights = norm.pdf(window, scale=sigma)\n",
    "\n",
    "    #Loop over enhancer rows in crispr dataframe\n",
    "    for ei, enh_row in crispr_df.iterrows() :\n",
    "        \n",
    "        #Get gene row in attribution score array\n",
    "        gi = gene_i[enh_row['gene']]\n",
    "        \n",
    "        #Get sequence start/end coordinates for attribution scores\n",
    "        enh_seq_starts[ei] = seq_starts[gi]\n",
    "        enh_seq_ends[ei] = seq_ends[gi]\n",
    "\n",
    "        #Extension boundaries\n",
    "        emid = (enh_row['start'] + enh_row['end']) // 2\n",
    "        enh_start = emid - elen_ext//2\n",
    "        enh_end = enh_start + elen_ext\n",
    "\n",
    "        #Map enhancer coordinates to gene sequence\n",
    "        eg_start = enh_start - enh_seq_starts[ei]\n",
    "        eg_end = enh_end - enh_seq_starts[ei]\n",
    "        eg_mid = emid - enh_seq_starts[ei]\n",
    "\n",
    "        #Check if enhancer is within sequence bounds\n",
    "        if eg_start < extra_padding :\n",
    "            pass #Enhancer outside of sequence\n",
    "        elif eg_end > seq_len - extra_padding :\n",
    "            pass #Enhancer outside of sequence\n",
    "        else :\n",
    "            \n",
    "            #Determine smoothing window (if gaussian smoothing is on)\n",
    "            if use_gaussian :\n",
    "                \n",
    "                eg_window = window + eg_mid\n",
    "                wstart, wend = 0, seq_len\n",
    "                \n",
    "                if eg_window[0] < 0:\n",
    "                    wstart = -eg_window[0]\n",
    "                if eg_window[-1] > seq_len:\n",
    "                    wend = seq_len - eg_window[-1] - 1\n",
    "                eg_window = eg_window[wstart:wend]\n",
    "          \n",
    "            #Mark enhancer as valid\n",
    "            enh_valids[ei] = 1\n",
    "\n",
    "            #Compute score (gaussian)\n",
    "            if use_gaussian :\n",
    "                enh_ref[ei] = np.sum(weights[wstart:wend] * np.abs(grads_ref[gi].astype('float32'))[eg_window]) / np.sum(weights[wstart:wend])\n",
    "            else : #Otherwise compute sum score\n",
    "                #Optionally use absolute value of attribution scores\n",
    "                if use_abs :\n",
    "                    enh_ref[ei] = np.abs(grads_ref[gi,eg_start:eg_end]).sum(dtype='float32')\n",
    "                else :\n",
    "                    enh_ref[ei] = grads_ref[gi,eg_start:eg_end].sum(dtype='float32')\n",
    "\n",
    "            #Optionally normalize scores\n",
    "            if normalize:\n",
    "                enh_ref[ei] /= np.abs(grads_ref[gi]).mean(dtype='float32')\n",
    "\n",
    "    return enh_ref, enh_seq_starts, enh_seq_ends, enh_valids\n",
    "\n",
    "#Helper function to remove gene id suffix\n",
    "def _trim_dot(gene_id):\n",
    "  \n",
    "    #Find suffix start\n",
    "    dot_i = gene_id.rfind('.')\n",
    "    \n",
    "    #Attempt to trim gene id\n",
    "    if dot_i == -1:\n",
    "        return gene_id\n",
    "    \n",
    "    return gene_id[:dot_i]\n",
    "\n",
    "#Function to compute attribution scores for crispr dataset\n",
    "def compute_scores_borzoi(satg_h5f, crispr_gtf_file, crispr_table_tsv, score_col='grads', normalize=True, use_abs=True, use_gaussian=False, sigma=300, tpm_file=None, min_tpm_rank=0., exons_gtf=None, allow_other_exons_to_overlap=True) :\n",
    "    \n",
    "    #Read sequences and reference attribution scores\n",
    "    with h5py.File(satg_h5f, 'r') as satg_h5 :\n",
    "        \n",
    "        #Get sequence start/.end coordinates\n",
    "        seq_starts = satg_h5['start'][:]\n",
    "        seq_ends = satg_h5['end'][:]\n",
    "\n",
    "        #Load gene ids and map gene id to row index\n",
    "        gene_ids = [_trim_dot(gene_id.decode('UTF-8')) for gene_id in satg_h5['gene']]\n",
    "        geneid_i = dict(zip(gene_ids, np.arange(len(gene_ids))))\n",
    "\n",
    "        #Initialize attribution score structure\n",
    "        num_seqs, seq_len, _, num_targets = satg_h5[score_col].shape\n",
    "        grads_ref = np.zeros((num_seqs,seq_len), dtype='float16')\n",
    "        \n",
    "        #Loop over sequences\n",
    "        for si in range(num_seqs) :\n",
    "\n",
    "            #Get scores\n",
    "            grads_si = satg_h5[score_col][si].sum(axis=-1)\n",
    "\n",
    "            #Get sequences and fill masked sequence positions\n",
    "            seq_1hot = satg_h5['seqs'][si]\n",
    "            n_mask = (seq_1hot.sum(axis=-1) == 0)\n",
    "            seq_1hot[n_mask] = np.array([1,0,0,0])\n",
    "            \n",
    "            #Slice scores\n",
    "            grads_ref[si] = grads_si[seq_1hot]\n",
    "\n",
    "    #Hash gene symbols to indexes\n",
    "    gene_i = {}\n",
    "    \n",
    "    #Loop over lines in crispr gtf file\n",
    "    for line in open(crispr_gtf_file) :\n",
    "        kv = gtf_kv(line.split('\\t')[8])\n",
    "        \n",
    "        #Get gene id base and symbol\n",
    "        gene_id = _trim_dot(kv['gene_id'])\n",
    "        gene_symbol = kv['gene_name']\n",
    "        \n",
    "        #Fill dictionaries\n",
    "        if gene_id in geneid_i:\n",
    "            gene_i[gene_symbol] = geneid_i[gene_id]\n",
    "\n",
    "    #Score crispr sites\n",
    "    crispr_df = pd.read_csv(crispr_table_tsv, sep='\\t')\n",
    "    crispr_df['score'], crispr_df['seq_start'], crispr_df['seq_end'], crispr_df['is_valid'] = _score_sites_borzoi(\n",
    "        crispr_df,\n",
    "        gene_i,\n",
    "        grads_ref,\n",
    "        seq_starts,\n",
    "        seq_ends,\n",
    "        normalize=normalize,\n",
    "        use_abs=use_abs,\n",
    "        use_gaussian=use_gaussian,\n",
    "        sigma=sigma\n",
    "    )\n",
    "\n",
    "    #Only keep sites within sequence window\n",
    "    crispr_df = crispr_df.loc[crispr_df['is_valid'] == 1].copy()\n",
    "\n",
    "    #Optionally load tpm rank dataframe for matched cell type\n",
    "    if tpm_file is not None :\n",
    "        tpm_df = pd.read_csv(tpm_file, sep='\\t')\n",
    "        \n",
    "        #Filter on minimum median TPM rank\n",
    "        crispr_df = crispr_df.join(tpm_df[['gene_symbol', 'rank_k562_median']].set_index('gene_symbol'), on='gene', how='inner')\n",
    "        crispr_df = crispr_df.loc[crispr_df['rank_k562_median'] >= min_tpm_rank].copy()\n",
    "  \n",
    "    #Optionally load exon annotation and remove crispr loci which overlap exons\n",
    "    if exons_gtf is not None :\n",
    "\n",
    "        #Read exons\n",
    "        transcriptome = bgene.Transcriptome(exons_gtf)\n",
    "\n",
    "        #Create bedtool object with exons\n",
    "        exons_bedt = transcriptome.bedtool_exon()\n",
    "\n",
    "        #Create bedtool object with crispr loci\n",
    "        crispr_bedt = _make_crispr_bedt(crispr_df, allow_other_exons_to_overlap)\n",
    "\n",
    "        #Optionally map gene id base to symbol\n",
    "        gene_symb = None\n",
    "        if allow_other_exons_to_overlap and tpm_df is not None :\n",
    "            gene_symb = {}\n",
    "            for _, row in tpm_df.iterrows() :\n",
    "                gene_symb[row['gene_base']] = row['gene_symbol']\n",
    "\n",
    "        #Intersect crispr loci with exon bed\n",
    "        overlapping_ids = []\n",
    "\n",
    "        #Loop over overlapping site-exon pairs\n",
    "        for overlap in exons_bedt.intersect(crispr_bedt, wo=True) :\n",
    "            exon_gene_id = _trim_dot(overlap[3])\n",
    "            enh_id = overlap[9]\n",
    "            enh_gene_symbol = overlap[10]\n",
    "\n",
    "            #Check if site is overlapping measured gene in crispr dataset\n",
    "            if gene_symb is not None and exon_gene_id in gene_symb and gene_symb[exon_gene_id] == enh_gene_symbol :\n",
    "                overlapping_ids.append(enh_id)\n",
    "            else :\n",
    "                overlapping_ids.append(enh_id)\n",
    "\n",
    "        #Get set of overlapping enhancer ids\n",
    "        overlapping_ids = list(set(overlapping_ids))\n",
    "\n",
    "        #Get enhancer id column\n",
    "        crispr_df['enh_id'] = crispr_df['chr'] + '_' + crispr_df['start'].astype(str) + '_' + crispr_df['end'].astype(str)\n",
    "        if allow_other_exons_to_overlap :\n",
    "            crispr_df['enh_id'] += '_' + crispr_df['gene'].astype(str)\n",
    "\n",
    "        #Keep non-overlapping crispr rows only\n",
    "        crispr_df = crispr_df.loc[~crispr_df['enh_id'].isin(overlapping_ids)].copy()\n",
    "    \n",
    "    return crispr_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b13cfeb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compute borzoi scores (ism_shuffle; ensemble)\n",
    "\n",
    "ablation_strs = [\n",
    "    'baseline',\n",
    "    'human_all',\n",
    "    'human_dnase_atac_rna',\n",
    "    'k562_all',\n",
    "    'k562_dnase_atac_rna',\n",
    "    'k562_rna',\n",
    "    'multisp_dnase_atac_rna',\n",
    "    'multisp_rna',\n",
    "    'multisp_no_unet',\n",
    "]\n",
    "\n",
    "for ablation_str in ablation_strs :\n",
    "\n",
    "    #Get crispr dataframe with scores\n",
    "    crispr_df = compute_scores_borzoi(\n",
    "        'flowfish_miborzoi_' + ablation_str + '_undo_ism_shuffle_clip/ism_mean.h5',\n",
    "        'crispr_genes.gtf',\n",
    "        'crispr_table.tsv',\n",
    "        score_col='isms',\n",
    "        normalize=False,\n",
    "        use_abs=False,\n",
    "        use_gaussian=False,\n",
    "        sigma=300,\n",
    "        tpm_file=None,\n",
    "        min_tpm_rank=0.,\n",
    "        exons_gtf=None,\n",
    "        allow_other_exons_to_overlap=True,\n",
    "    )\n",
    "\n",
    "    #Compute percent changes from log fold changes\n",
    "    crispr_df['pct_change'] = (np.exp(-crispr_df['score']) - 1) * 100\n",
    "    crispr_df['score'] = np.abs(crispr_df['score'])\n",
    "\n",
    "    #Store stats\n",
    "    crispr_df.to_csv(\"site_scores_flowfish_k562_ism_miborzoi_\" + ablation_str + \"_ensemble.csv\", sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b0f6d8cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Compute borzoi scores (ism_shuffle; cross-fold)\n",
    "\n",
    "ablation_strs = [\n",
    "    'baseline',\n",
    "    'human_all',\n",
    "    'human_dnase_atac_rna',\n",
    "    'k562_all',\n",
    "    'k562_dnase_atac_rna',\n",
    "    'k562_rna',\n",
    "    'multisp_dnase_atac_rna',\n",
    "    'multisp_rna',\n",
    "    'multisp_no_unet',\n",
    "]\n",
    "\n",
    "fold_index = [0, 1]\n",
    "\n",
    "for ablation_str in ablation_strs :\n",
    "\n",
    "    for fold_ix in fold_index :\n",
    "    \n",
    "        #Get crispr dataframe with scores\n",
    "        crispr_df = compute_scores_borzoi(\n",
    "            'flowfish_miborzoi_' + ablation_str + '_undo_ism_shuffle_clip/ism_f' + str(fold_ix) + 'c0.h5',\n",
    "            'crispr_genes.gtf',\n",
    "            'crispr_table.tsv',\n",
    "            score_col='isms',\n",
    "            normalize=False,\n",
    "            use_abs=False,\n",
    "            use_gaussian=False,\n",
    "            sigma=300,\n",
    "            tpm_file=None,\n",
    "            min_tpm_rank=0.,\n",
    "            exons_gtf=None,\n",
    "            allow_other_exons_to_overlap=True,\n",
    "        )\n",
    "\n",
    "        #Compute percent changes from log fold changes\n",
    "        crispr_df['pct_change'] = (np.exp(-crispr_df['score']) - 1) * 100\n",
    "        crispr_df['score'] = np.abs(crispr_df['score'])\n",
    "\n",
    "        #Store stats\n",
    "        crispr_df.to_csv(\"site_scores_flowfish_k562_ism_miborzoi_\" + ablation_str + \"_f\" + str(fold_ix) + \".csv\", sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "735a3b00",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
