{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d49dce53",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import h5py\n",
    "\n",
    "from sklearn.metrics import average_precision_score, roc_auc_score\n",
    "\n",
    "#Function to do a number of permuted average precision calculations\n",
    "def permute_ap_auc(df_snp_pos_in, df_snp_neg_in, frac_sample=0.8, n_shuffles=1000) :\n",
    "    \n",
    "    df_snp_pos = df_snp_pos_in.copy()\n",
    "    df_snp_neg = df_snp_neg_in.copy()\n",
    "    \n",
    "    df_snp_pos['is_causal'] = 1\n",
    "    df_snp_neg['is_causal'] = 0\n",
    "\n",
    "    shuffle_index_pos = np.arange(len(df_snp_pos))\n",
    "\n",
    "    aps_borzoi = []\n",
    "    aps_saluki = []\n",
    "\n",
    "    aucs_borzoi = []\n",
    "    aucs_saluki = []\n",
    "\n",
    "    for shuffle_ix in range(n_shuffles) :\n",
    "\n",
    "        np.random.shuffle(shuffle_index_pos)\n",
    "        df_snp_pos_shuffled = df_snp_pos.iloc[shuffle_index_pos].copy().reset_index(drop=True)\n",
    "\n",
    "        #Sample positives\n",
    "        df_snp_pos_shuffled = df_snp_pos_shuffled.iloc[:int(frac_sample * len(df_snp_pos_shuffled))].copy().reset_index(drop=True)\n",
    "        \n",
    "        #Select matched negatives\n",
    "        df_snp_neg_shuffled = df_snp_neg.loc[df_snp_neg['matched_snp'].isin(df_snp_pos_shuffled['matched_snp'])].copy().reset_index(drop=True)\n",
    "\n",
    "        df_snp = pd.concat([df_snp_pos_shuffled, df_snp_neg_shuffled])\n",
    "\n",
    "        score_borzoi = np.abs(np.array(df_snp['delta_score_borzoi'].values))\n",
    "        score_saluki = np.abs(np.array(df_snp['SSD0'].values))\n",
    "        label = np.array(df_snp['is_causal'].values)\n",
    "\n",
    "        #Precision-recall scores\n",
    "        ap_borzoi = average_precision_score(label, score_borzoi)\n",
    "        ap_saluki = average_precision_score(label, score_saluki)\n",
    "        \n",
    "        aps_borzoi.append(ap_borzoi)\n",
    "        aps_saluki.append(ap_saluki)\n",
    "\n",
    "        #Area under ROC curve\n",
    "        auc_borzoi = roc_auc_score(label, score_borzoi)\n",
    "        auc_saluki = roc_auc_score(label, score_saluki)\n",
    "        \n",
    "        aucs_borzoi.append(auc_borzoi)\n",
    "        aucs_saluki.append(auc_saluki)\n",
    "    \n",
    "    ap_bundle = [\n",
    "        [np.quantile(aps_borzoi, q=0.10), np.mean(aps_borzoi), np.quantile(aps_borzoi, q=0.90)],\n",
    "        [np.quantile(aps_saluki, q=0.10), np.mean(aps_saluki), np.quantile(aps_saluki, q=0.90)]\n",
    "    ]\n",
    "    auc_bundle = [\n",
    "        [np.quantile(aucs_borzoi, q=0.10), np.mean(aucs_borzoi), np.quantile(aucs_borzoi, q=0.90)],\n",
    "        [np.quantile(aucs_saluki, q=0.10), np.mean(aucs_saluki), np.quantile(aucs_saluki, q=0.90)]\n",
    "    ]\n",
    "    \n",
    "    return [ap_bundle, auc_bundle]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "05b4d48a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inflating predictions for 'pos'.\n",
      "len(df) = 944\n",
      "len(df) = 944\n",
      "Inflating predictions for 'neg'.\n",
      "len(df) = 7550\n",
      "len(df) = 7550\n",
      "\n",
      "len(len(df_snp_pos)) = 46\n",
      "len(len(df_snp_neg)) = 350\n",
      "\n",
      "ap_borzoi  = 0.243\n",
      "ap_saluki  = 0.220\n",
      "\n",
      "auc_borzoi = 0.654\n",
      "auc_saluki = 0.668\n",
      "\n",
      "ap_borzoi_mean  = 0.245 [0.208, 0.280]\n",
      "ap_saluki_mean  = 0.225 [0.198, 0.255]\n",
      "\n",
      "auc_borzoi_mean = 0.654 [0.626, 0.684]\n",
      "auc_saluki_mean = 0.667 [0.640, 0.694]\n"
     ]
    }
   ],
   "source": [
    "#Inflate pos predictions\n",
    "\n",
    "def inflate_snp_predictions(pos_or_neg, fold_index=[0, 1], fold_suffixes=['', '']) :\n",
    "\n",
    "    print(\"Inflating predictions for '\" + pos_or_neg + \"'.\")\n",
    "    \n",
    "    df_folds = []\n",
    "    \n",
    "    sed_scores = None\n",
    "\n",
    "    for fold_i, [fold_ix, fold_suffix] in enumerate(zip(fold_index, fold_suffixes)) :\n",
    "\n",
    "        sed = h5py.File('models/f' + str(fold_ix) + 'c0' + fold_suffix + '/complex_3prime/merge_' + pos_or_neg + '/sed.h5', 'r')\n",
    "        \n",
    "        gene_id_mat = sed['gene'][()]\n",
    "        snp_mat = sed['snp'][()]\n",
    "        chr_mat = sed['chr'][()]\n",
    "        pos_mat = sed['pos'][()]\n",
    "        score_mat = sed['logSED'][()]\n",
    "        ref_allele_mat = sed['ref_allele'][()]\n",
    "        alt_allele_mat = sed['alt_allele'][()]\n",
    "        \n",
    "        if fold_i == 0 :\n",
    "            sed_scores = np.abs(score_mat.copy())\n",
    "        else :\n",
    "            sed_scores += np.abs(score_mat.copy())\n",
    "        \n",
    "        gene_ids = []\n",
    "        snps = []\n",
    "        chroms = []\n",
    "        poses = []\n",
    "        preds = []\n",
    "        ref_alleles = []\n",
    "        alt_alleles = []\n",
    "\n",
    "        for i, si in enumerate(sed['si'][()].tolist()) :\n",
    "\n",
    "            gene_ids.append(gene_id_mat[i].decode())\n",
    "            snps.append(snp_mat[si].decode())\n",
    "            chroms.append(chr_mat[si].decode())\n",
    "            poses.append(int(pos_mat[si]))\n",
    "\n",
    "            preds.append(np.mean(np.abs(score_mat[i, :])))\n",
    "\n",
    "            ref_alleles.append(ref_allele_mat[si].decode())\n",
    "            alt_alleles.append(alt_allele_mat[si].decode())\n",
    "\n",
    "        df = pd.DataFrame({\n",
    "            'gene_id_pred' : gene_ids,\n",
    "            'snp' : snps,\n",
    "            'chrom' : chroms,\n",
    "            'pos' : poses,\n",
    "            'delta_score' : preds,\n",
    "        })\n",
    "\n",
    "        print(\"len(df) = \" + str(len(df)))\n",
    "\n",
    "        df_folds.append(df)\n",
    "\n",
    "    df = df_folds[0]\n",
    "    df = df.rename(columns={'delta_score' : 'delta_score_0'})\n",
    "    df['delta_score'] = df['delta_score_0']\n",
    "    \n",
    "    for i in range(1, len(df_folds)) :\n",
    "\n",
    "        df['delta_score_' + str(i)] = df_folds[i]['delta_score']\n",
    "        df['delta_score'] += df['delta_score_' + str(i)]\n",
    "        \n",
    "    df['delta_score'] /= float(len(df_folds))\n",
    "    \n",
    "    sed_scores /= float(len(df_folds))\n",
    "    \n",
    "    df['row_index'] = np.arange(len(df), dtype=np.int32)\n",
    "    \n",
    "    #Get finemapped gene id and distance\n",
    "    vcf_df = pd.read_csv('/home/jlinder/seqnn/data/traitgym/complex_3prime/' + pos_or_neg + '_merge.vcf', sep='\\t', skiprows=3)\n",
    "    \n",
    "    vcf_df['min_dist'] = vcf_df['INFO'].apply(lambda x: x.split(\"DI=\")[1].split(\";\")[0]).astype(int)\n",
    "    vcf_df['matched_snp'] = vcf_df['INFO'].apply(lambda x: x.split(\"MG=\")[1].split(\";\")[0])\n",
    "    \n",
    "    df = df.join(vcf_df[['ID', 'min_dist', 'matched_snp']].set_index('ID'), on='snp', how='inner').copy().reset_index(drop=True)\n",
    "    \n",
    "    df.loc[df['delta_score'].isnull(), 'delta_score'] = 0.\n",
    "    \n",
    "    return df, sed_scores\n",
    "\n",
    "df_snp_pos, scores_pos = inflate_snp_predictions('pos', [0, 1], ['', ''])\n",
    "df_snp_neg, scores_neg = inflate_snp_predictions('neg', [0, 1], ['', ''])\n",
    "\n",
    "#Rename borzoi VEP score\n",
    "df_snp_pos = df_snp_pos.rename(columns={'delta_score' : 'delta_score_borzoi'})\n",
    "df_snp_neg = df_snp_neg.rename(columns={'delta_score' : 'delta_score_borzoi'})\n",
    "\n",
    "#Get target ids\n",
    "target_ids = h5py.File('models/f0c0/gtexg/merge_pos/sed.h5', 'r')['target_ids'][()]\n",
    "target_ids = [tid.decode() for tid in target_ids.tolist()]\n",
    "\n",
    "target_name_dict = pd.read_csv('targets_human_rna.txt', sep='\\t')[['identifier', 'description']].set_index('identifier').to_dict(orient='index')\n",
    "\n",
    "target_names = []\n",
    "for tid in target_ids :\n",
    "    target_names.append(target_name_dict[tid]['description'])\n",
    "\n",
    "#Filter on matched gene\n",
    "'''\n",
    "df_snp_pos['gene_id_base'] = df_snp_pos['gene_id_pred'].apply(lambda x: x.split(\".\")[0])\n",
    "df_snp_neg['gene_id_base'] = df_snp_neg['gene_id_pred'].apply(lambda x: x.split(\".\")[0])\n",
    "\n",
    "df_snp_pos = df_snp_pos.query(\"gene_id_base == finemap_gene_id\").copy().reset_index(drop=True)\n",
    "df_snp_neg = df_snp_neg.query(\"gene_id_base == finemap_gene_id\").copy().reset_index(drop=True)\n",
    "'''\n",
    "\n",
    "#Filter on max score\n",
    "df_snp_pos = df_snp_pos.sort_values(by=['delta_score_borzoi'], ascending=False).drop_duplicates(subset=['snp'], keep='first').copy().reset_index(drop=True)\n",
    "df_snp_neg = df_snp_neg.sort_values(by=['delta_score_borzoi'], ascending=False).drop_duplicates(subset=['snp'], keep='first').copy().reset_index(drop=True)\n",
    "\n",
    "scores_pos = scores_pos[df_snp_pos['row_index'].values.tolist(), :]\n",
    "scores_neg = scores_neg[df_snp_neg['row_index'].values.tolist(), :]\n",
    "\n",
    "#Append saluki ensemble predictions\n",
    "\n",
    "df_saluki_pos = None\n",
    "df_saluki_neg = None\n",
    "for fold_ix in range(10) :\n",
    "    for cross_ix in range(5) :\n",
    "        df_saluki_pos_curr = pd.read_csv('/home/jlinder/analysis/saluki/complex_3prime/f' + str(fold_ix) + 'c' + str(cross_ix) + '/pos_merge/ssd.tsv', sep='\\t')\n",
    "        df_saluki_neg_curr = pd.read_csv('/home/jlinder/analysis/saluki/complex_3prime/f' + str(fold_ix) + 'c' + str(cross_ix) + '/neg_merge/ssd.tsv', sep='\\t')\n",
    "        \n",
    "        df_saluki_pos_curr['SSD0'] = np.abs(df_saluki_pos_curr['SSD0'])\n",
    "        df_saluki_neg_curr['SSD0'] = np.abs(df_saluki_neg_curr['SSD0'])\n",
    "        \n",
    "        if df_saluki_pos is None :\n",
    "            df_saluki_pos = df_saluki_pos_curr.copy()\n",
    "            df_saluki_neg = df_saluki_neg_curr.copy()\n",
    "        else :\n",
    "            df_saluki_pos['SSD0'] += df_saluki_pos_curr['SSD0'] / 50.\n",
    "            df_saluki_neg['SSD0'] += df_saluki_neg_curr['SSD0'] / 50.\n",
    "\n",
    "#df_saluki_pos['SSD0'] = np.abs(df_saluki_pos['SSD0'])\n",
    "#df_saluki_neg['SSD0'] = np.abs(df_saluki_neg['SSD0'])\n",
    "\n",
    "#Drop duplicate variants\n",
    "df_saluki_pos = df_saluki_pos.groupby('variant').agg({'SSD0' : 'max'}).copy().reset_index()\n",
    "df_saluki_neg = df_saluki_neg.groupby('variant').agg({'SSD0' : 'max'}).copy().reset_index()\n",
    "\n",
    "df_snp_pos = df_snp_pos.join(df_saluki_pos.set_index('variant'), on='snp', how='inner').copy().reset_index(drop=True)\n",
    "df_snp_neg = df_snp_neg.join(df_saluki_neg.set_index('variant'), on='snp', how='inner').copy().reset_index(drop=True)\n",
    "\n",
    "#Filter negatives without matching positive\n",
    "df_snp_neg = df_snp_neg.loc[df_snp_neg['matched_snp'].isin(df_snp_pos['matched_snp'])].copy().reset_index(drop=True)\n",
    "\n",
    "#Sort by SNP\n",
    "df_snp_pos = df_snp_pos.sort_values(by=['snp'], ascending=True).copy().reset_index(drop=True)\n",
    "df_snp_neg = df_snp_neg.sort_values(by=['snp'], ascending=True).copy().reset_index(drop=True)\n",
    "\n",
    "print(\"\")\n",
    "print(\"len(len(df_snp_pos)) = \" + str(len(df_snp_pos)))\n",
    "print(\"len(len(df_snp_neg)) = \" + str(len(df_snp_neg)))\n",
    "\n",
    "#Evaluate SNP classification performance based on scalar effect size predictions\n",
    "\n",
    "df_snp_pos['is_causal'] = 1\n",
    "df_snp_neg['is_causal'] = 0\n",
    "df_snp = pd.concat([df_snp_pos, df_snp_neg])\n",
    "\n",
    "score_borzoi = np.abs(np.array(df_snp['delta_score_borzoi'].values))\n",
    "score_saluki = np.abs(np.array(df_snp['SSD0'].values))\n",
    "label = np.array(df_snp['is_causal'].values)\n",
    "\n",
    "#Precision-recall scores\n",
    "ap_borzoi = average_precision_score(label, score_borzoi)\n",
    "ap_saluki = average_precision_score(label, score_saluki)\n",
    "\n",
    "#Area under ROC curve\n",
    "auc_borzoi = roc_auc_score(label, score_borzoi)\n",
    "auc_saluki = roc_auc_score(label, score_saluki)\n",
    "\n",
    "print(\"\")\n",
    "print('ap_borzoi  = %.3f' % ap_borzoi)\n",
    "print('ap_saluki  = %.3f' % ap_saluki)\n",
    "\n",
    "print(\"\")\n",
    "print('auc_borzoi = %.3f' % auc_borzoi)\n",
    "print('auc_saluki = %.3f' % auc_saluki)\n",
    "\n",
    "#Evaluate SNP classification performance based on scalar effect size predictions, permutation\n",
    "\n",
    "ap_bundle, auc_bundle = permute_ap_auc(df_snp_pos, df_snp_neg)\n",
    "\n",
    "[ap_borzoi_lo, ap_borzoi_mean, ap_borzoi_hi], [ap_saluki_lo, ap_saluki_mean, ap_saluki_hi] = ap_bundle\n",
    "[auc_borzoi_lo, auc_borzoi_mean, auc_borzoi_hi], [auc_saluki_lo, auc_saluki_mean, auc_saluki_hi] = auc_bundle\n",
    "\n",
    "print(\"\")\n",
    "print(('ap_borzoi_mean  = %.3f' % ap_borzoi_mean) + (' [%.3f' % ap_borzoi_lo) + (', %.3f' % ap_borzoi_hi) + ']')\n",
    "print(('ap_saluki_mean  = %.3f' % ap_saluki_mean) + (' [%.3f' % ap_saluki_lo) + (', %.3f' % ap_saluki_hi) + ']')\n",
    "\n",
    "print(\"\")\n",
    "print(('auc_borzoi_mean = %.3f' % auc_borzoi_mean) + (' [%.3f' % auc_borzoi_lo) + (', %.3f' % auc_borzoi_hi) + ']')\n",
    "print(('auc_saluki_mean = %.3f' % auc_saluki_mean) + (' [%.3f' % auc_saluki_lo) + (', %.3f' % auc_saluki_hi) + ']')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "56278ad7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inflating predictions for 'pos'.\n",
      "len(df) = 637\n",
      "len(df) = 637\n",
      "Inflating predictions for 'neg'.\n",
      "len(df) = 4283\n",
      "len(df) = 4283\n",
      "\n",
      "len(len(df_snp_pos)) = 29\n",
      "len(len(df_snp_neg)) = 216\n",
      "\n",
      "ap_borzoi  = 0.485\n",
      "ap_saluki  = 0.418\n",
      "\n",
      "auc_borzoi = 0.785\n",
      "auc_saluki = 0.791\n",
      "\n",
      "ap_borzoi_mean  = 0.489 [0.436, 0.540]\n",
      "ap_saluki_mean  = 0.423 [0.367, 0.480]\n",
      "\n",
      "auc_borzoi_mean = 0.787 [0.756, 0.820]\n",
      "auc_saluki_mean = 0.793 [0.759, 0.831]\n"
     ]
    }
   ],
   "source": [
    "#Inflate pos predictions\n",
    "\n",
    "def inflate_snp_predictions(pos_or_neg, fold_index=[0, 1], fold_suffixes=['', '']) :\n",
    "\n",
    "    print(\"Inflating predictions for '\" + pos_or_neg + \"'.\")\n",
    "    \n",
    "    df_folds = []\n",
    "    \n",
    "    sed_scores = None\n",
    "\n",
    "    for fold_i, [fold_ix, fold_suffix] in enumerate(zip(fold_index, fold_suffixes)) :\n",
    "\n",
    "        sed = h5py.File('models/f' + str(fold_ix) + 'c0' + fold_suffix + '/mendelian_3prime/merge_' + pos_or_neg + '/sed.h5', 'r')\n",
    "        \n",
    "        gene_id_mat = sed['gene'][()]\n",
    "        snp_mat = sed['snp'][()]\n",
    "        chr_mat = sed['chr'][()]\n",
    "        pos_mat = sed['pos'][()]\n",
    "        score_mat = sed['logSED'][()]\n",
    "        ref_allele_mat = sed['ref_allele'][()]\n",
    "        alt_allele_mat = sed['alt_allele'][()]\n",
    "        \n",
    "        if fold_i == 0 :\n",
    "            sed_scores = np.abs(score_mat.copy())\n",
    "        else :\n",
    "            sed_scores += np.abs(score_mat.copy())\n",
    "        \n",
    "        gene_ids = []\n",
    "        snps = []\n",
    "        chroms = []\n",
    "        poses = []\n",
    "        preds = []\n",
    "        ref_alleles = []\n",
    "        alt_alleles = []\n",
    "\n",
    "        for i, si in enumerate(sed['si'][()].tolist()) :\n",
    "\n",
    "            gene_ids.append(gene_id_mat[i].decode())\n",
    "            snps.append(snp_mat[si].decode())\n",
    "            chroms.append(chr_mat[si].decode())\n",
    "            poses.append(int(pos_mat[si]))\n",
    "\n",
    "            preds.append(np.mean(np.abs(score_mat[i, :])))\n",
    "\n",
    "            ref_alleles.append(ref_allele_mat[si].decode())\n",
    "            alt_alleles.append(alt_allele_mat[si].decode())\n",
    "\n",
    "        df = pd.DataFrame({\n",
    "            'gene_id_pred' : gene_ids,\n",
    "            'snp' : snps,\n",
    "            'chrom' : chroms,\n",
    "            'pos' : poses,\n",
    "            'delta_score' : preds,\n",
    "        })\n",
    "\n",
    "        print(\"len(df) = \" + str(len(df)))\n",
    "\n",
    "        df_folds.append(df)\n",
    "\n",
    "    df = df_folds[0]\n",
    "    df = df.rename(columns={'delta_score' : 'delta_score_0'})\n",
    "    df['delta_score'] = df['delta_score_0']\n",
    "    \n",
    "    for i in range(1, len(df_folds)) :\n",
    "\n",
    "        df['delta_score_' + str(i)] = df_folds[i]['delta_score']\n",
    "        df['delta_score'] += df['delta_score_' + str(i)]\n",
    "        \n",
    "    df['delta_score'] /= float(len(df_folds))\n",
    "    \n",
    "    sed_scores /= float(len(df_folds))\n",
    "    \n",
    "    df['row_index'] = np.arange(len(df), dtype=np.int32)\n",
    "    \n",
    "    #Get finemapped gene id and distance\n",
    "    vcf_df = pd.read_csv('/home/jlinder/seqnn/data/traitgym/mendelian_3prime/' + pos_or_neg + '_merge.vcf', sep='\\t', skiprows=3)\n",
    "    \n",
    "    vcf_df['min_dist'] = vcf_df['INFO'].apply(lambda x: x.split(\"DI=\")[1].split(\";\")[0]).astype(int)\n",
    "    vcf_df['matched_snp'] = vcf_df['INFO'].apply(lambda x: x.split(\"MG=\")[1].split(\";\")[0])\n",
    "    \n",
    "    df = df.join(vcf_df[['ID', 'min_dist', 'matched_snp']].set_index('ID'), on='snp', how='inner').copy().reset_index(drop=True)\n",
    "    \n",
    "    df.loc[df['delta_score'].isnull(), 'delta_score'] = 0.\n",
    "    \n",
    "    return df, sed_scores\n",
    "\n",
    "df_snp_pos, scores_pos = inflate_snp_predictions('pos', [0, 1], ['', ''])\n",
    "df_snp_neg, scores_neg = inflate_snp_predictions('neg', [0, 1], ['', ''])\n",
    "\n",
    "#Rename borzoi VEP score\n",
    "df_snp_pos = df_snp_pos.rename(columns={'delta_score' : 'delta_score_borzoi'})\n",
    "df_snp_neg = df_snp_neg.rename(columns={'delta_score' : 'delta_score_borzoi'})\n",
    "\n",
    "#Get target ids\n",
    "target_ids = h5py.File('models/f0c0/gtexg/merge_pos/sed.h5', 'r')['target_ids'][()]\n",
    "target_ids = [tid.decode() for tid in target_ids.tolist()]\n",
    "\n",
    "target_name_dict = pd.read_csv('targets_human_rna.txt', sep='\\t')[['identifier', 'description']].set_index('identifier').to_dict(orient='index')\n",
    "\n",
    "target_names = []\n",
    "for tid in target_ids :\n",
    "    target_names.append(target_name_dict[tid]['description'])\n",
    "\n",
    "#Filter on matched gene\n",
    "'''\n",
    "df_snp_pos['gene_id_base'] = df_snp_pos['gene_id_pred'].apply(lambda x: x.split(\".\")[0])\n",
    "df_snp_neg['gene_id_base'] = df_snp_neg['gene_id_pred'].apply(lambda x: x.split(\".\")[0])\n",
    "\n",
    "df_snp_pos = df_snp_pos.query(\"gene_id_base == finemap_gene_id\").copy().reset_index(drop=True)\n",
    "df_snp_neg = df_snp_neg.query(\"gene_id_base == finemap_gene_id\").copy().reset_index(drop=True)\n",
    "'''\n",
    "\n",
    "#Filter on max score\n",
    "df_snp_pos = df_snp_pos.sort_values(by=['delta_score_borzoi'], ascending=False).drop_duplicates(subset=['snp'], keep='first').copy().reset_index(drop=True)\n",
    "df_snp_neg = df_snp_neg.sort_values(by=['delta_score_borzoi'], ascending=False).drop_duplicates(subset=['snp'], keep='first').copy().reset_index(drop=True)\n",
    "\n",
    "scores_pos = scores_pos[df_snp_pos['row_index'].values.tolist(), :]\n",
    "scores_neg = scores_neg[df_snp_neg['row_index'].values.tolist(), :]\n",
    "\n",
    "#Append saluki ensemble predictions\n",
    "\n",
    "df_saluki_pos = None\n",
    "df_saluki_neg = None\n",
    "for fold_ix in range(10) :\n",
    "    for cross_ix in range(5) :\n",
    "        df_saluki_pos_curr = pd.read_csv('/home/jlinder/analysis/saluki/mendelian_3prime/f' + str(fold_ix) + 'c' + str(cross_ix) + '/pos_merge/ssd.tsv', sep='\\t')\n",
    "        df_saluki_neg_curr = pd.read_csv('/home/jlinder/analysis/saluki/mendelian_3prime/f' + str(fold_ix) + 'c' + str(cross_ix) + '/neg_merge/ssd.tsv', sep='\\t')\n",
    "        \n",
    "        df_saluki_pos_curr['SSD0'] = np.abs(df_saluki_pos_curr['SSD0'])\n",
    "        df_saluki_neg_curr['SSD0'] = np.abs(df_saluki_neg_curr['SSD0'])\n",
    "        \n",
    "        if df_saluki_pos is None :\n",
    "            df_saluki_pos = df_saluki_pos_curr.copy()\n",
    "            df_saluki_neg = df_saluki_neg_curr.copy()\n",
    "        else :\n",
    "            df_saluki_pos['SSD0'] += df_saluki_pos_curr['SSD0'] / 50.\n",
    "            df_saluki_neg['SSD0'] += df_saluki_neg_curr['SSD0'] / 50.\n",
    "\n",
    "#df_saluki_pos['SSD0'] = np.abs(df_saluki_pos['SSD0'])\n",
    "#df_saluki_neg['SSD0'] = np.abs(df_saluki_neg['SSD0'])\n",
    "\n",
    "#Drop duplicate variants\n",
    "df_saluki_pos = df_saluki_pos.groupby('variant').agg({'SSD0' : 'max'}).copy().reset_index()\n",
    "df_saluki_neg = df_saluki_neg.groupby('variant').agg({'SSD0' : 'max'}).copy().reset_index()\n",
    "\n",
    "df_snp_pos = df_snp_pos.join(df_saluki_pos.set_index('variant'), on='snp', how='inner').copy().reset_index(drop=True)\n",
    "df_snp_neg = df_snp_neg.join(df_saluki_neg.set_index('variant'), on='snp', how='inner').copy().reset_index(drop=True)\n",
    "\n",
    "#Filter negatives without matching positive\n",
    "df_snp_neg = df_snp_neg.loc[df_snp_neg['matched_snp'].isin(df_snp_pos['matched_snp'])].copy().reset_index(drop=True)\n",
    "\n",
    "#Sort by SNP\n",
    "df_snp_pos = df_snp_pos.sort_values(by=['snp'], ascending=True).copy().reset_index(drop=True)\n",
    "df_snp_neg = df_snp_neg.sort_values(by=['snp'], ascending=True).copy().reset_index(drop=True)\n",
    "\n",
    "print(\"\")\n",
    "print(\"len(len(df_snp_pos)) = \" + str(len(df_snp_pos)))\n",
    "print(\"len(len(df_snp_neg)) = \" + str(len(df_snp_neg)))\n",
    "\n",
    "#Evaluate SNP classification performance based on scalar effect size predictions\n",
    "\n",
    "df_snp_pos['is_causal'] = 1\n",
    "df_snp_neg['is_causal'] = 0\n",
    "df_snp = pd.concat([df_snp_pos, df_snp_neg])\n",
    "\n",
    "score_borzoi = np.abs(np.array(df_snp['delta_score_borzoi'].values))\n",
    "score_saluki = np.abs(np.array(df_snp['SSD0'].values))\n",
    "label = np.array(df_snp['is_causal'].values)\n",
    "\n",
    "#Precision-recall scores\n",
    "ap_borzoi = average_precision_score(label, score_borzoi)\n",
    "ap_saluki = average_precision_score(label, score_saluki)\n",
    "\n",
    "#Area under ROC curve\n",
    "auc_borzoi = roc_auc_score(label, score_borzoi)\n",
    "auc_saluki = roc_auc_score(label, score_saluki)\n",
    "\n",
    "print(\"\")\n",
    "print('ap_borzoi  = %.3f' % ap_borzoi)\n",
    "print('ap_saluki  = %.3f' % ap_saluki)\n",
    "\n",
    "print(\"\")\n",
    "print('auc_borzoi = %.3f' % auc_borzoi)\n",
    "print('auc_saluki = %.3f' % auc_saluki)\n",
    "\n",
    "#Evaluate SNP classification performance based on scalar effect size predictions, permutation\n",
    "\n",
    "ap_bundle, auc_bundle = permute_ap_auc(df_snp_pos, df_snp_neg)\n",
    "\n",
    "[ap_borzoi_lo, ap_borzoi_mean, ap_borzoi_hi], [ap_saluki_lo, ap_saluki_mean, ap_saluki_hi] = ap_bundle\n",
    "[auc_borzoi_lo, auc_borzoi_mean, auc_borzoi_hi], [auc_saluki_lo, auc_saluki_mean, auc_saluki_hi] = auc_bundle\n",
    "\n",
    "print(\"\")\n",
    "print(('ap_borzoi_mean  = %.3f' % ap_borzoi_mean) + (' [%.3f' % ap_borzoi_lo) + (', %.3f' % ap_borzoi_hi) + ']')\n",
    "print(('ap_saluki_mean  = %.3f' % ap_saluki_mean) + (' [%.3f' % ap_saluki_lo) + (', %.3f' % ap_saluki_hi) + ']')\n",
    "\n",
    "print(\"\")\n",
    "print(('auc_borzoi_mean = %.3f' % auc_borzoi_mean) + (' [%.3f' % auc_borzoi_lo) + (', %.3f' % auc_borzoi_hi) + ']')\n",
    "print(('auc_saluki_mean = %.3f' % auc_saluki_mean) + (' [%.3f' % auc_saluki_lo) + (', %.3f' % auc_saluki_hi) + ']')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9daa33a0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
